### PR TITLE
use checkbox icons for staging and unstaging

### PR DESCRIPTION
**Pull request by Murmele**
Instead of those arrows which might be unclear to understand

### CODE REVIEW

**Comment by Murmele** [comment]
One pixel row too much below the checkbox
image

**Comment by Murmele** [comment]
https://user-images.githubusercontent.com/10099533/150636675-12b19852-2306-4e9a-9e28-b85302147ab6.mp4

**Comment by mmahmoudian** [comment]
Yes I think this looks good now. Well done!

**Comment by stefanknotzer** [comment]
Oh boy the Windows build looks awful...
Give me a few days to fix that.

**Comment by Murmele** [comment]
How does it look?😄 why it looks differently?

File: src/editor/TextEditor.cpp [hunk 1]
```diff
@@ -16,11 +16,23 @@
 #include <QStyle>
 #include <QWindow>
```
Comment (stefanknotzer) [inline src/editor/TextEditor.cpp:17]:
```suggestion
#include <QStyle>
#include <QStyleFactory>
#include <QWindow>
```

File: src/editor/TextEditor.cpp [hunk 2]
```diff
@@ -16,11 +16,23 @@
...
+namespace {
+QPixmap getStagedUnstagedIcon(bool staged) {
+    QCheckBox cb;
+    cb.setChecked(staged);
+    // size must be restricted, otherwise the last row contains 2
+    // pixels of the background color and the others have only
+    // one which looks ugly
+    return cb.grab(QRect(QPoint(0,0), QSize(14, 14)));
+}
+}
```
Comment (stefanknotzer) [inline src/editor/TextEditor.cpp:34]:
```suggestion
namespace {

QPixmap stagedUnstagedIcon(
  const bool &checked,
  const QColor &background,
  const int &fontHeight)
{
  // Set background color and checkbox size.
  QString checkBoxStyle =
    "QCheckBox {"
    "  background: %1;"
    "}"
    "QCheckBox::indicator {"
    "  height: %2;"
    "  width: %2;"
    "}";

  QCheckBox checkBox;
  checkBox.setChecked(checked);
#if defined(Q_OS_WIN)
  checkBox.setStyle(QStyleFactory::create("windows"));
#endif
  checkBox.setStyleSheet(checkBoxStyle.arg(background.name()).arg(fontHeight - 2));
  return checkBox.grab(QRect(QPoint(0,0), QSize(fontHeight, fontHeight)));
}

} // anon. namespace
```

File: src/editor/TextEditor.cpp [hunk 3]
```diff
@@ -38,7 +50,8 @@ TextEditor::TextEditor(QWidget *parent)
   mNoteIcon = style->standardIcon(QStyle::SP_MessageBoxInformation);
   mWarningIcon = style->standardIcon(QStyle::SP_MessageBoxWarning);
   mErrorIcon = style->standardIcon(QStyle::SP_MessageBoxCritical);
-  mStagedIcon = style->standardIcon(QStyle::SP_ArrowUp);
+  mStagedIcon = getStagedUnstagedIcon(true);
+  mUnStagedIcon = getStagedUnstagedIcon(false);
```
Comment (stefanknotzer) [inline src/editor/TextEditor.cpp:54]:
```suggestion
```

File: src/editor/TextEditor.cpp [hunk 4]
```diff
@@ -186,7 +200,8 @@ void TextEditor::applySettings()
   loadMarkerIcon(ErrorMarker, mErrorIcon);
 
 
-  loadMarkerIcon(StagedMarker, mStagedIcon);
+  loadMarkerPixmap(StagedMarker, mStagedIcon);
+  loadMarkerPixmap(UnstagedMarker, mUnStagedIcon);
```
Comment (stefanknotzer) [inline src/editor/TextEditor.cpp:204]:
```suggestion
  // Initialize markers.
  QColor background = palette().color(QPalette::Base);
  int fontHeight = textHeight(0);
  mStagedIcon = stagedUnstagedIcon(
    true, background, fontHeight);
  mUnStagedIcon = stagedUnstagedIcon(
    false, background, fontHeight);
  if (mStatusDiff)
    setMarginWidthN(Staged, fontHeight);
```

File: src/editor/TextEditor.h [hunk 5]
```diff
@@ -162,6 +162,7 @@ class TextEditor : public Scintilla::ScintillaIFace
 private:
   int diagnosticMarker(int line);
   void loadMarkerIcon(Marker marker, const QIcon &icon);
+  void loadMarkerPixmap(Marker marker, const QPixmap &icon);
```
Comment (stefanknotzer) [inline src/editor/TextEditor.h:165]:
```suggestion
  void loadMarkerPixmap(Marker marker, const QPixmap &pixmap);
```

**Comment by Murmele** [comment]
@stefanknotzer 
There is again this problem with this thick border on the bottom and on the right side.
Additionally I prefer the size of the Checkboxes smaller, so they look like the checkboxes of the filewidgets.
What do you think?
image

File: src/editor/TextEditor.cpp [hunk 6]
```diff
@@ -14,13 +14,43 @@
...
+    "  width: %2;"
+    "}";
+
+  QCheckBox checkBox;
+  checkBox.setChecked(checked);
+#if defined(Q_OS_WIN)
+  checkBox.setStyle(QStyleFactory::create("windows"));
+#endif
+  checkBox.setStyleSheet(checkBoxStyle.arg(background.name()).arg(fontHeight - 2));
+  return checkBox.grab(QRect(QPoint(0,0), QSize(fontHeight, fontHeight)));
```
Comment (stefanknotzer) [inline src/editor/TextEditor.cpp:49]:
```suggestion
  return checkBox.grab(QRect(QPoint(0,0), QSize(fontHeight - 2, fontHeight - 2)));
```

Comment (stefanknotzer) [reply src/editor/TextEditor.cpp:49]: My fault: I reduced the checkbox size by 2 to get 2 pixel margin between checkboxes in subsequent lines.
But finally I forgot to reduce the size in the grab() command...

The checkbox works fine with Linux and Windows and all 3 themes. To get rid of the checkbox border/shadow you would need to draw the pixmap with rectangles or lines always minding the themed brush and background color.
A fixed size checkbox could look strange with different font sizes - so why not scaling the checkbox to the editor font size?

**Comment by Murmele** [comment]
But should we then resize also the Filewidget Checkbox? Otherwise it looks quite ugly or what do you think?

**Comment by stefanknotzer** [comment]
I am not a designer and I like the look as it is.

My personal intention is to use the ui elements as they are. There are way to many stylesheets and other ui manipulations in the code. This might make the ui beautiful in someones eyes but for me the function comes first. Always. And the more stylesheets and resizes we code into the source the harder it is to maintain.

TLDR; I am the wrong guy for that kind of questions :grin:

**Comment by Martin2278** [comment]
grafik
Does it looks for you the same?

**Comment by stefanknotzer** [comment]
@Martin2278 
Windows? If true: I already had my doubts and think I should ditch the whole 
```
#if defined(Q_OS_WIN)
  checkBox.setStyle(QStyleFactory::create("windows"));
#endif
```
Then the indicator looks weird at some special sizes but the style is right.

**Comment by Martin2278** [comment]
@stefanknotzer can you show how it looks like if you remove this define?

**Comment by stefanknotzer** [comment]
@Martin2278 
Theme: Dark, Courier New 10:
Dark_10

Theme: System, Courier New, 7:
System_7

Theme: System, Courier New 10:
System_10
(the indicator look is a bit off)

Theme: System, Courier New 16:
System_16
(the indicator is pixelated - but someone using that font size probably can't see that 😁)

**Comment by Martin2278** [comment]
Are we able to increase the checkbox size before exporting the pixmap before. Will it change anything?

File: src/editor/TextEditor.cpp [hunk 7]
```diff
@@ -14,13 +14,43 @@
 #include <QMainWindow>
 #include <QScrollBar>
 #include <QStyle>
+#include <QStyleFactory>
```
Comment (stefanknotzer) [inline src/editor/TextEditor.cpp:17]:
```suggestion

```

**Comment by Murmele** [comment]
@stefanknotzer 
This looks in the flatpak package when scrolling in the editor and resizing the text. How do you change the size?
image

**Comment by stefanknotzer** [comment]
The `fontPointSize()` is not looking good on (my) linux and Windows (wrong size, blurry border).
I also think you are working around a Scintilla bug.
My suggestion to get this into main:
```
#if defined(FLATPAK)
  .. = fontPointSize(0) * ..;
#else
  .. = textHeight(0);
#endif
```

**Comment by Murmele** [comment]
Yes, then we should also write a comment there so that it is clear, or at least linking to this mr

**Comment by Murmele** [comment]
@stefanknotzer Can you have a look on the latest commit and merge if you are fine with it?

**Review by stefanknotzer — APPROVED** [review]
